service/controller/resource/workermigration/create.go

```diff
@@ -0,0 +1,505 @@
+	_, err = r.ensureSparkExists(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured Spark CR exists for legacy workers VMSS")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding if node pool workers are ready")
+	if !machinePool.Status.InfrastructureReady || (machinePool.Status.Replicas != machinePool.Status.ReadyReplicas) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "node pool workers are not ready yet")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+		return nil
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "found that node pool workers are ready")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring that legacy workers have drainerconfig cr")
+	err = r.ensureDrainerConfigsExists(ctx, azureAPI, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured that legacy workers have drainerconfig cr")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring that legacy workers are drained")
+	allNodesDrained, err := r.allDrainerConfigsWithDrainedState(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !allNodesDrained {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "some old worker nodes are still draining")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "deleting timed out drainerconfigs")
+
+		// In case some draining operations timed out, delete CRs so that they
+		// are recreated on next reconciliation.
+		err = r.deleteTimedOutDrainerConfigs(ctx, cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "deleted timed out drainerconfigs")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+		return nil
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured that legacy workers are drained")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleting drainerconfigs")
+	err = r.deleteDrainerConfigs(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted drainerconfigs")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleting legacy workers' deployment")
+	err = azureAPI.DeleteDeployment(ctx, key.ResourceGroupName(cr), key.WorkersVmssDeploymentName)
+	if azure.IsNotFound(err) {
+		// It's ok. We would have deleted it anyway ¯\_(ツ)_/¯
+	} else if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted legacy workers' deployment")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleting legacy workers' VMSS")
+	err = azureAPI.DeleteVMSS(ctx, key.ResourceGroupName(cr), *legacyVMSS.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted legacy workers' VMSS")
```

Comment:
what's deleting the `AzureConfig` once the migration is done?

Reply:
Nothing yet, I hope. We reconcile the tenant cluster control plane by it. :slightly_smiling_face: